fdbclient/NativeAPI.actor.cpp

```diff
@@ -1096,10 +1101,15 @@ Future<Void> DatabaseContext::onMasterProxiesChanged() {
 	return this->masterProxiesChangeTrigger.onTrigger();
 }
 
-bool DatabaseContext::sampleReadTags() {
+bool DatabaseContext::sampleReadTags() const {
 	return clientInfo->get().transactionTagSampleRate > 0 && deterministicRandom()->random01() <= clientInfo->get().transactionTagSampleRate;
 }
 
+bool DatabaseContext::sampleOnCost(uint64_t cost) const {
+	if(clientInfo->get().transactionTagSampleCost <= 0) return false;
+	return deterministicRandom()->random01() <= (double)cost / clientInfo->get().transactionTagSampleCost;
```

Comment:
Yes. I experimented that even every transaction commits with this struct, the effect on commit latency is negligible.

Reply:
Does it have any effect on client throughput? Clients are often limited in how much work they can do by the fact that there's a single network thread, so that might be a good thing to check too.

Reply:
I checked the mako benchmark using `./bin/mako --cluster fdb.cluster --mode run --rows 1000000 --transaction "gu100i100"` to compare attach tag to every transaction and no tag on transactions. In each benchmark, I configure from a brand-new database. The TPSs are 45.60 (tagged) and 43.99 (no tag). 

Since here all transactions do 200 writes that let them always be sampled, I think the cost is negligible.

Reply:
Is this workload saturating the client network thread?

Reply:
I think the network workload is also negligible. By check `ProcessMetrics` trace event, the `MainThreadCPUSeconds` is about 1.85~2.68 when always attach tags while `MainThreadCPUSeconds` is about 1.99~2.71 without any tags.

Reply:
Mainly I'm looking to get a handle on what the throughput looks like when saturating the network thread, which would mean running a workload that fully utilizes it and then measuring how many transactions we get done.

Reply:
Can we merge this and see whether your circus nightlies find a regression? Our regression tests don't find anything, and it is kind of hard to come up with a special benchmark just for this... If what you're saying is indeed a problem, we can still revert or optimize this.

Reply:
I'm not sure we have a test for this in Circus either, and I'm also a little bit skeptical that the current process around Circus would have us detecting an issue and reporting it in a timely manner.

Reply:
I guess we do have a few binding throughput tests and an RYW throughput test that might be able to meet this need, but I still have my second concern.

Reply:
So what would you suggest we do? Xiaoxi ran benchmarks but couldn't find regressions and the whole thing can pretty much be deactivated through a knob. Also the next release is not imminent - so in my opinion all necessary precautions have been met. So what would you suggest we do?

Reply:
Run the test described above, with a client workload that is heavy enough to be saturating on the client side. Then measure the throughput of the client. If mako is able to specify a higher transaction rate, etc., then setting that high enough should be sufficient.

Reply:
If mako can't already do it, then I think you're probably right. In that case we can go ahead and merge and defer this specific test. Let me know if that's the case.

Reply:
Mako kind of can do that - but the bottleneck is typically some latches to ship tasks to the main thread (and I would assume this is the case for most workloads with many threads). Kao and Xiaoxi are working on running this workload for now - but I am not convinced this will address your concern.

I think to test this we would need to run a workload that is much harder to run where the client generates mostly large transactions with many threads and the underlying cluster is large enough to not be saturated by this

Reply:
Ok, I see. We can see how this test goes, but if it doesn't successfully saturate I'll go ahead and merge anyway.

Reply:
The test results are here:
* setting: 144 threads, 100 write/txn, `MainThreadCpuSeconds/Elapsed` is about 4.94/5.0
* no tag: TPS for 3 tests are 469, 485, 472
* always with tag: TPS for 3 tests are 473, 472, 473
With this setting, every transaction has a tag and cost estimation sent with its commit request.

Reply:
Ok, thank you for running that. I don’t think there are any other blockers now, so I can merge it.